dbms/src/DataStreams/ParallelInputsProcessor.h

```diff
@@ -285,9 +286,12 @@ class ParallelInputsProcessor
                 std::lock_guard lock(available_inputs_mutex);
 
                 /// If there are no free sources, then this thread is no longer needed. (But other threads can work with their sources.)
-                if (available_inputs.empty())
+                if (0 == may_available_inputs)
```

Comment:
Sleep a while before next try?

Reply:
Or use condition variable since sleep will introduce extra random latency for queries, like `SharedQueryBlockInputStream`, refer to https://github.com/pingcap/tiflash/issues/4473 for details.

Reply:
how about using `MPMCQueue`?